src/NuGet.Core/NuGet.Commands/IncludeFlagUtils.cs

```diff
@@ -51,98 +81,70 @@ internal static class IncludeFlagUtils
+            // already correct. If the existing node has less flags then the
+            // walk must continue and all new flags found combined with the
+            // existing ones.
+            while (nodeQueue.Count > 0)
             {
-                directDependencies.AddRange(specFramework.Dependencies);
-            }
+                node = nodeQueue.Dequeue();
+                var rootId = node.Item.Key.Name;
 
-            foreach (var dependency in directDependencies)
-            {
-                if (result.ContainsKey(dependency.Name))
+                // Combine results on the way up
+                LibraryIncludeFlags currentTypes;
+                if (result.TryGetValue(rootId, out currentTypes))
                 {
-                    result[dependency.Name] = dependency.IncludeType;
+                    if ((node.DependencyType & currentTypes) == node.DependencyType)
+                    {
+                        // Noop, this is done
+                        continue;
+                    }
+
+                    result[rootId] = (currentTypes | node.DependencyType);
                 }
                 else
                 {
-                    result.Add(dependency.Name, dependency.IncludeType);
-                }
-            }
-
-            return result;
-        }
-
-        private static void FlattenDependencyTypesUnified(
-            Dictionary<string, LibraryIncludeType> result,
-            GraphItem<RemoteResolveResult> root,
-            GraphEdge<RemoteResolveResult> outerEdge,
-            Dictionary<string, GraphItem<RemoteResolveResult>> unifiedNodes,
-            LibraryIncludeType dependencyType)
-        {
-            var rootId = root.Key.Name;
-
-            var hasCycle = false;
-
-            var cursor = outerEdge;
-            while (cursor != null)
-            {
-                if (rootId.Equals(cursor.Item.Key.Name, StringComparison.OrdinalIgnoreCase))
-                {
-                    // Cycle detected
-                    hasCycle = true;
-                    break;
+                    result.Add(rootId, node.DependencyType);
                 }
 
-                cursor = cursor.OuterEdge;
-            }
-
-            if (!hasCycle)
-            {
-                // Intersect on the way down
-                foreach (var dependency in root.Data.Dependencies)
+                foreach (var dependency in node.Item.Data.Dependencies)
                 {
                     // Any nodes that are not in unifiedNodes are types that should be ignored
                     // We should also ignore dependencies that are excluded
                     GraphItem<RemoteResolveResult> child;
                     if (unifiedNodes.TryGetValue(dependency.Name, out child)
-                        && !dependency.SuppressParent.Equals(LibraryIncludeType.All))
+                        && dependency.SuppressParent != LibraryIncludeFlags.All)
```

Comment:
I like this idea and logically it's completely correct, but there is more going on here. If (which wouldn't happen) a package ended up with no flags at all then we still need to mark the dependencies as having no flags in the result dict. The suppress parent=all scenario is special because that is caught during the dependency resolution and those dependencies are ignored there and not in the graph (unless they came in through another route, but they still shouldn't be followed here).